src/category_theory/monad/monadicity.lean

```diff
@@ -0,0 +1,372 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+
+import category_theory.limits.shapes.reflexive
+import category_theory.limits.preserves.limits
+import category_theory.monad.limits
+import category_theory.monad.coequalizer
+
+/-!
+# Monadicity theorems
+
+We prove monadicity theorems which can establish a given functor is monadic. In particular, we
+show three versions of Beck's monadicity theorem, and the reflexive (crude) monadicity theorem:
+
+`G` is a monadic right adjoint if it has a right adjoint, and:
+
+* `D` has, `G` preserves and reflects `G`-split coequalizers, see
+  `category_theory.monad.monadic_of_has_preserves_reflects_G_split_coequalizers`
+* `G` creates `G`-split coequalizers, see
+  `category_theory.monad.monadic_of_creates_G_split_coequalizers`
+  (The converse of this is also shown, see
+   `category_theory.monad.creates_G_split_coequalizers_of_monadic`)
+* `D` has and `G` preserves `G`-split coequalizers, and `G` reflects isomorphisms, see
+  `category_theory.monad.monadic_of_has_preserves_G_split_coequalizers_of_reflects_isomorphisms`
+* `D` has and `G` preserves reflexive coequalizers, and `G` reflects isomorphisms, see
+  `category_theory.monad.monadic_of_has_preserves_reflexive_coequalizers_of_reflects_isomorphisms`
+
+## Tags
+
+Beck, monadicity, descent
+
+## TODO
+
+Dualise to show comonadicity theorems.
+-/
+universes v₁ v₂ u₁ u₂
+
+namespace category_theory
+namespace monad
+open limits
+
+noncomputable theory
+-- Hide the implementation details in this namespace.
+namespace monadicity_internal
+
+section
+
+-- We use these parameters and abbreviations to simplify the statements of internal constructions
+-- here.
+parameters {C : Type u₁} {D : Type u₂}
+parameters [category.{v₁} C] [category.{v₁} D]
+parameters {G : D ⥤ C} [is_right_adjoint G]
+
+/-- An internal convenience abbreviation to make lemma statements clearer. -/
+abbreviation F : C ⥤ D := left_adjoint G
```

Comment:
Can you just use local notation instead? Because abbreviations will persist outside of this file.

Reply:
I tried this first but then I can't use `F.obj`, it has to be `(F).obj` or something annoyingly :( 
Would private abbreviation be okay?

Reply:
No, I think that would only make it worse :wink: 
My point is that people who want use these results (or results that use those results) might end up using `comparison_adjunction_unit_f` (or some other lemma from this file) and end up with `F` in their tactic state. And they will think: huh, what is `F`. (The docstring for `F` will not help them either.)
Outside of this file, I would prefer to just have `left_adjoint G`. I understand that `(F).obj` is annoying, but maybe it's worth the hassle. Or maybe ``local notation `la` := left_adjoint``, or something like that?

Reply:
Alright, I'll switch to `(F).obj`. The idea was that people using this file wouldn't need to touch anything other than the main theorems at the end, much like the cartesian closed presheaves stuff from before

Reply:
Removed all the abbreviations! The statements look so weird now, but oh well :)